File: appmonet-sdk/appmonet-sdk-dfp/src/main/java/com/monet/bidder/DFPAdViewRequest.java
package com.monet.bidder;

import android.location.Location;
import android.os.Bundle;
import com.google.ads.mediation.admob.AdMobAdapter;
import com.google.android.gms.ads.AdRequest;
import com.google.android.gms.ads.doubleclick.PublisherAdRequest;
import com.google.android.gms.ads.mediation.MediationAdRequest;
import com.google.android.gms.ads.mediation.admob.AdMobExtras;
import com.monet.bidder.auction.AuctionRequest;
import java.util.Date;

class DFPAdViewRequest extends AdServerAdRequest {
  private static AdRequest mAdRequest;

  DFPAdViewRequest() {
    mAdRequest = new AdRequest.Builder().build();
  }

  DFPAdViewRequest(AdRequest adRequest) {
    mAdRequest = adRequest;
  }

  DFPAdViewRequest(MediationAdRequest mediationAdRequest) {
    mAdRequest = new AdRequest.Builder()
        .setBirthday(mediationAdRequest.getBirthday())
        .setGender(mediationAdRequest.getGender())
        .setLocation(mediationAdRequest.getLocation())
        .build();
  }

  private Bundle getAdMobExtras() {
    try {
      Bundle extras = mAdRequest.getNetworkExtrasBundle(AdMobAdapter.class);
      //      AdMobExtras extras = mAdRequest.getNetworkExtras(AdMobExtras.class);
      if (extras != null) {
        return extras;
      }
    } catch (Exception e) {
    }

    return new Bundle();
  }

  @Override
  public Bundle getCustomTargeting() {
    // also get the admob extras and merge it here
    Bundle extras = getAdMobExtras();

    // create a bundle merging both
    Bundle merged = new Bundle();
    merged.putAll(extras);

    return merged;
  }

  @Override
  public Date getBirthday() {
    return mAdRequest.getBirthday();
  }

  @Override
  public String getGender() {
    switch (mAdRequest.getGender()) {
      case PublisherAdRequest.GENDER_FEMALE:
        return "female";
      case PublisherAdRequest.GENDER_MALE:
        return "male";
      default:
        return "unknown";
    }
  }

  @Override
  public Location getLocation() {
    return mAdRequest.getLocation();
  }

  @Override
  public String getContentUrl() {
    return mAdRequest.getContentUrl();
  }

  AdRequest getDFPRequest() {
    return mAdRequest;
  }

  @Override
  public AuctionRequest apply(AuctionRequest request, AdServerAdView adView) {
    // transfer admob extras if they're there
    try {
      Bundle adMob = mAdRequest.getNetworkExtrasBundle(AdMobAdapter.class);
      AdMobExtras admobExtras = mAdRequest.getNetworkExtras(AdMobExtras.class);
      request.getAdmobExtras().putAll(filterTargeting(
          (admobExtras != null) ? admobExtras.getExtras() : adMob));
    } catch (Exception e) {
      // do nothing
    }

    if (request.getRequestData() == null) {
      request.setRequestData(new RequestData(this, adView));
    }
    return request;
  }

  @Override
  public String getPublisherProvidedId() {
    return "";
  }

  static DFPAdViewRequest fromAuctionRequest(AuctionRequest request) {
    AdRequest.Builder builder = new AdRequest.Builder()
        .addCustomEventExtrasBundle(CustomEventBanner.class, request.getNetworkExtras())
        .addCustomEventExtrasBundle(CustomEventInterstitial.class, request.getNetworkExtras())
        .addCustomEventExtrasBundle(MonetDfpCustomEventInterstitial.class,
            request.getNetworkExtras())
        .addNetworkExtrasBundle(CustomEventBanner.class, request.getNetworkExtras());

    // add in the admob extras
    Bundle completeExtras =
        request.getAdmobExtras() == null ? new Bundle() : request.getAdmobExtras();
    completeExtras.putAll(request.getTargeting());
    try {
      builder.addNetworkExtrasBundle(AdMobAdapter.class, completeExtras);
    } catch (Exception e) {
      sLogger.error("excetion " + e);
      // do nothing
    }
    if (request.getRequestData() != null) {
      if (request.getRequestData().contentURL != null
          && !request.getRequestData().contentURL.isEmpty()) {
        builder.setContentUrl(request.getRequestData().contentURL);
      }
      builder.setLocation(request.getRequestData().location);
    }
    AdRequest adRequest = builder.build();
    return new DFPAdViewRequest(adRequest);
  }
}

File: appmonet-sdk/appmonet-sdk-dfp/src/main/java/com/monet/bidder/DFPAdViewRequest.kt
package com.monet.bidder

import android.location.Location
import android.os.Bundle
import com.google.ads.mediation.admob.AdMobAdapter
import com.google.android.gms.ads.AdRequest
import com.google.android.gms.ads.AdRequest.Builder
import com.google.android.gms.ads.doubleclick.PublisherAdRequest
import com.google.android.gms.ads.mediation.MediationAdRequest
import com.google.android.gms.ads.mediation.admob.AdMobExtras
import com.monet.bidder.auction.AuctionRequest
import com.monet.bidder.bid.BidResponse
import java.util.Date

internal class DFPAdViewRequest : AdServerAdRequest {
  val dfpRequest: AdRequest

  constructor() {
    dfpRequest = Builder().build()
  }

  constructor(adRequest: AdRequest) {
    dfpRequest = adRequest
  }

  constructor(mediationAdRequest: MediationAdRequest) {
    dfpRequest = Builder()
        .setBirthday(mediationAdRequest.birthday)
        .setGender(mediationAdRequest.gender)
        .setLocation(mediationAdRequest.location)
        .build()
  }

  private val adMobExtras: Bundle
    get() {
      try {
        val extras = dfpRequest.getNetworkExtrasBundle(
            AdMobAdapter::class.java
        )
        if (extras != null) {
          return extras
        }
      } catch (e: Exception) {
        //do nothing
      }
      return Bundle()
    }
  override val customTargeting: Bundle
    get() {
      val extras = adMobExtras
      // create a bundle merging both
      val merged = Bundle()
      merged.putAll(extras)
      return merged
    }

  override val birthday: Date?
    get() = dfpRequest.birthday

  override val gender: String?
    get() {
      return when (dfpRequest.gender) {
        PublisherAdRequest.GENDER_FEMALE -> "female"
        PublisherAdRequest.GENDER_MALE -> "male"
        else -> "unknown"
      }
    }
  override val bid: BidResponse?
    get() = null

  override val location: Location?
    get() = dfpRequest.location

  override val contentUrl: String?
    get() = dfpRequest.contentUrl

  override fun apply(
    request: AuctionRequest,
    adView: AdServerAdView
  ): AuctionRequest {
    // transfer admob extras if they're there
    try {
      val adMob = dfpRequest.getNetworkExtrasBundle(
          AdMobAdapter::class.java
      )
      val adMobExtras = dfpRequest.getNetworkExtras(
          AdMobExtras::class.java
      )
      request.admobExtras.putAll(
          filterTargeting(
              if (adMobExtras != null) adMobExtras.extras else adMob
          )
      )
    } catch (e: Exception) {
      // do nothing
    }
    if (request.requestData == null) {
      request.requestData = RequestData(this, adView)
    }
    return request
  }

  override val publisherProvidedId: String?
    get() = null

  companion object {
    fun fromAuctionRequest(request: AuctionRequest): DFPAdViewRequest {
      val builder = Builder()
          .addCustomEventExtrasBundle(CustomEventBanner::class.java, request.networkExtras)
          .addCustomEventExtrasBundle(CustomEventInterstitial::class.java, request.networkExtras)
          .addCustomEventExtrasBundle(
              MonetDfpCustomEventInterstitial::class.java,
              request.networkExtras
          )
          .addNetworkExtrasBundle(CustomEventBanner::class.java, request.networkExtras)

      // add in the admob extras
      val completeExtras = request.admobExtras
      completeExtras.putAll(request.targeting)
      try {
        builder.addNetworkExtrasBundle(AdMobAdapter::class.java, completeExtras)
      } catch (e: Exception) {
        sLogger.error("excetion $e")
        // do nothing
      }
      if (request.requestData != null) {
        if (!request.requestData!!.contentURL.isNullOrEmpty()) {
          builder.setContentUrl(request.requestData!!.contentURL)
        }
        builder.setLocation(request.requestData!!.location)
      }
      val adRequest = builder.build()
      return DFPAdViewRequest(adRequest)
    }
  }
}